scanner/matcher/updater/updater.go

```diff
@@ -257,6 +265,12 @@ func (u *Updater) runUpdate(ctx context.Context) error {
 		Str("timestamp", prevTimestamp.Format(http.TimeFormat)).
 		Msg("previous vuln update")
 
+	// If this is not the first updater run among all instances, we assume we are
+	// initialized.
+	if !prevTimestamp.IsZero() && u.initialized.CompareAndSwap(false, true) {
```

Comment:
I think there may be an issue. Let's say there are two matcher replicas, A and B. Let's say A always happens to be the lucky one who handles the vuln update cycles. It's possible B always fails fast after calling `TryLock`. In that case, B may never acknowledge it's initialized.

I think we should read the last update timestamp before calling `TryLock` (this should be fine, as the lock is just meant to ensure only one replica updates vulns).

Also note for B, it will only do this check every 5 minutes (by default). This is probably ok, though. 5 minutes isn't terribly long in the grand scheme of things

Reply:
Ah, I thought the updater thread would hold waiting, instead of failing. What is the rationale for failing instead of making all updaters wait for a chance to run? That would allow another updater instance to retry immediately if one of its peers fails instead of waiting for the next run. Are you concerned with something here?

Reply:
This was essentially taken from ClairCore: https://github.com/quay/claircore/blob/v1.5.21/libvuln/updates/manager.go#L210

The idea is even if the updater can't run now, it'll just try again soon after (about 5 minutes, by default), so there isn't a huge need to have each updater wait to attempt immediately after the current one succeeds/fails. The ones which fail to obtain the lock don't necessarily fail. We log `Msg("did not obtain lock, skipping update run")`, as this is normal behavior

Reply:
By "failing", I mean "give up".

Reply:
We need to move the check before the try-lock then.

Reply:
The problem with your suggestion is that it's possible that one instance reads the zero timestamp just before the other updater is successful. This instance might fail the lock and remain "not ready" for the next update time. It might be enough time to block traffic to this pod.